sdk/data/azcosmos/cosmos_location_cache.go

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcosmos
+
+import (
+	"net/url"
+	"sync"
+	"time"
+)
+
+const defaultExpirationTime time.Duration = time.Minute * 5
+
+const (
+	none opType = iota
+	read
+	write
+	all
+)
+
+type opType int
+
+type locationUnavailabilityInfo struct {
+	lastCheckTime  time.Time
+	unavailableOps opType
+}
+
+type dbAcctLocationsInfo struct {
+	prefLocations              []string
+	availWriteLocations        []string
+	availReadLocations         []string
+	availWriteEndptsByLocation map[string]url.URL
+	availReadEndptsByLocation  map[string]url.URL
+	writeEndpts                []url.URL
+	readEndpts                 []url.URL
+}
+
+// AcctRegion represents a Azure Cosmos DB Database Account in a specific region.
+type AcctRegion struct {
+	// The name of the database account location in the Azure Cosmos DB service.
+	name string
+	// The URL of the database account location in the Azure Cosmos DB service.
+	endpoint string
+}
+
+// AcctProperties represents a container for Azure Cosmos DB databases.
+type AcctProperties struct {
+	// The list of readable regions of this database account from the Azure Cosmos DB service.
+	readRegions []AcctRegion
+	// The list of writable regions of this database account from the Azure Cosmos DB service.
+	writeRegions []AcctRegion
+	// Indicates whether multiple write locations are enabled for this database account.
+	enableMultipleWriteLocations bool
+}
+
+type locationCache struct {
+	locationInfo              dbAcctLocationsInfo
+	defaultEndpt              url.URL
+	enableEndptDiscovery      bool
+	useMultipleWriteLocations bool
+	rwMutex                           sync.RWMutex
+	locationUnavailabilityInfoMap     map[url.URL]locationUnavailabilityInfo
```

Comment:
Should this be a `sync.Map` ? https://pkg.go.dev/sync#Map

Reply:
I used the mutex + normal map combination to deal with some things that a sync.Map doesn't support, such as getting the length/checking if non-empty. I can restructure this to use a sync.Map and figure out other workarounds, but the underlying functionality is the same.

Reply:
I just thought that using a language provided construct versus fabricating one ourselves would be better. Besides obtaining the length (which I think is not exactly that, you need to just know if there is at least 1 item, right?) which are the other gaps?

Reply:
The other major reason that I preferred to utilize this approach is to have more control over it. I believe for sync.map, you cannot limit the key/value types, i.e. you can put anything into it. Do you think we should still change it?